mlops blog post for deploying a model to sagemaker

usrbinkat opened the pull request:
## Description

TODO:

- [ ] Figma meta image

## Checklist:

- [x] I have reviewed the [style guide](https://github.com/pulumi/pulumi-hugo/blob/master/STYLE-GUIDE.md).
- [x] If blogging, I have reviewed the [blogging guide](https://github.com/pulumi/pulumi-hugo/blob/master/BLOGGING.md).
- [x] I have manually confirmed that all new links work.
- [ ] I added aliases (i.e., redirects) for all filename changes.
- [ ] If making css changes, I rebuilt the bundle.

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+---
+date: 2023-09-08
+title: "Deploy a Huggingface Model on AWS SageMaker with Pulumi Python IaC"
```
Is it `Huggingface` (one word) or `Hugging Face` (two words)? I've seen it both ways. The template, for example, has it as two words.

usrbinkat replied:
I looked it up when scott mentioned it during template review. The corp is `Hugging Face Inc.` however they are also routinely marketed as either the emoji or `Huggingface`. Given the corporate name I'm okay with committing to Hugging Face and we can make huggingface all together the meta tag.

scottslowe replied:
Fixed throughout in 46fb9a7

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+---
+date: 2023-09-08
+title: "Deploy a Huggingface Model on AWS SageMaker with Pulumi Python IaC"
```
`s/AWS Sagemaker/Amazon Sagemaker/g`

scottslowe replied:
Fixed via 46fb9a7

aaronkao commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+    - huggingface
+---
+
+Welcome to another post in the Pulumi + Python MLOps #MLOpsChallenge series!
```
I really don't like the hashtag. Its not meant to be a challenge. In the previous post I called it "AI Infrastructure as Python". We can turn that into a hashtag if we want.

aaronkao replied:
Also instead of focusing on the fact its a series, I would explain the concept or the problem we are solving. "Deploying a HF model on Sagemaker takes a bit of setup to spin up. Engineers run into these XYZ problems, here is how Pulumi solves it for you"

This way, you have a nice intro to why they should bother to try this out

elistaykova replied:
+1. Can we move away from the MLops challenge branding, which then spills into social posts and so one? The whole concept gets messy.

We are trying to accelerate engineers spinning up AI infra. It's pretty simple.

Also, i am very keen on the post explaining the concept and problem you are solving with this: It's challenging to do X, because of this and this. Here is a magical way to solve it. 
Then you show the solution and outcome

scottslowe replied:
I've taken the liberty of adding a quick draft on why this is important and why readers should care. @elistaykova & @aaronkao, please review and provide feedback.

usrbinkat replied:
Can we move this type of conversation back to the campaign planning doc and associated meetings? Feedback extending beyond a single blog post content review is best facilitated on the planning end of content creation.

usrbinkat replied:
proceeding with hashtag #PulumiMLOps per zoom call decision

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+date: 2023-09-08
+title: "Deploy a Huggingface Model on AWS SageMaker with Pulumi Python IaC"
+allow_long_title: true
+meta_desc: "Guided short tutorial on starting a Pulumi Infrastructure as Code project to deploy Huggingface LLM Models on Amazon Web Services SageMaker Machine Learning Platform"
```
```suggestion
meta_desc: "Guided short tutorial on starting a Pulumi infrastructure as code project to deploy Huggingface LLM Models on Amazon SageMaker machine learning platform"
```

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+
+Welcome to another post in the Pulumi + Python MLOps #MLOpsChallenge series!
+
+In this short tutorial we will deploy a publicly available Huggingface model on Amazon AWS SageMaker, and test a prompt using a short python script.
```
```suggestion
In this short tutorial we will deploy a publicly available Huggingface model on Amazon SageMaker, and test a prompt using a short Python script.
```

aaronkao replied:
```suggestion
In this short tutorial we will deploy a publicly available Huggingface model on Amazon AWS SageMaker, and we will test a prompt using a short python script.
```

aaronkao commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+
+Welcome to another post in the Pulumi + Python MLOps #MLOpsChallenge series!
+
+In this short tutorial we will deploy a publicly available Huggingface model on Amazon AWS SageMaker, and test a prompt using a short python script.
```
I believe their official title is Amazon Sagemaker, but pick one

scottslowe replied:
Fixed throughout in 46fb9a7

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+* [Amazon AWS SageMaker Model Endpoint](https://www.pulumi.com/registry/packages/aws/api-docs/sagemaker/model/)
+* [Amazon AWS CloudWatch MetricAlarm](https://www.pulumi.com/registry/packages/aws/api-docs/cloudwatch/metricalarm/)
+
+[Pulumi templates](../how-to-create-and-share-a-pulumi-template) enhance Infrastructure as Code (IaC) development by providing a pre-written code framework designed to help you bootstrap new pulumi projects quickly.
```
```suggestion
[Pulumi templates](../how-to-create-and-share-a-pulumi-template) enhance the infrastructure as code (IaC) development experience by providing a pre-written code framework designed to help you bootstrap new Pulumi projects quickly.
```

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+
+[Pulumi templates](../how-to-create-and-share-a-pulumi-template) enhance Infrastructure as Code (IaC) development by providing a pre-written code framework designed to help you bootstrap new pulumi projects quickly.
+
+You can use the `sagemaker-aws-python` template as a working python starting point for your own AWS SageMaker deployments, and customize the model you use, SageMaker configuration, and CloudWatch integration just to scratch the surface. Your only limits are your imagination.
```
I'll stop pointing out AWS SageMaker from here on, but please make sure you catch all instances.

scottslowe replied:
Fixed via 46fb9a7

aaronkao commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
````diff
@@ -0,0 +1,131 @@
+pulumi login
+```
+
+### 2. Prepare a new Pulumi Project
````
isn't there a `pulumi config set aws:region <REGION>` somewhere in here. Yours goes direct to us-east-1, but I think you have to set it first right?

usrbinkat replied:
it's a default provided by the template, user's are prompted to accept the default or override with their desired region when creating the template.

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+
+You can use the `sagemaker-aws-python` template as a working python starting point for your own AWS SageMaker deployments, and customize the model you use, SageMaker configuration, and CloudWatch integration just to scratch the surface. Your only limits are your imagination.
+
+## Requirements
```
Before getting into the nitty-gritty details, can we add some high-level text about why this is important? You know, to give someone some context about the business value of deploying an HF LLM to SageMaker? That'll help ground this and provide some justification for _why_ someone should care about this post.

aaronkao replied:
Agreed, its about what is the challenge and difficulty of doing this the alternative/normal way. Then why Pulumi provides an easier solution

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+
+### 2. Prepare a new Pulumi Project
+
+Here we create our new pulumi project directory and populate it from the `sagemaker-aws-python` [Pulumi template](https://github.com/pulumi/templates)
```
```suggestion
Here we create our new Pulumi project directory and populate it from the `sagemaker-aws-python` [Pulumi template](https://github.com/pulumi/templates)
```

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+
+### 3. Deploy your model as a new SageMaker endpoint
+
+This step may take between 10 and 20 minutes while Amazon AWS builds your infrastructure. You can follow along in the console as resources are provisioned, or open the link displayed in terminal for your stack in a browser to follow along there too!
```
```suggestion
This step may take between 10 and 20 minutes while Amazon builds your infrastructure. You can follow along in the console as resources are provisioned, or open the link displayed in terminal for your stack in a browser to follow along there too!
```

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+
+Once your stack has finished deploying, use this rudimentary Python snippet to test the deployed SageMaker endpoint.
+
+> NOTE: notice that we are using `us-east-1` in this script, be sure to change the reagion in python to match the reagion you deployed the SageMaker endpoint into.
```
```suggestion
> NOTE: notice that we are using `us-east-1` in this script, be sure to change the reagion in python to match the region you deployed the SageMaker endpoint into.
```

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+
+### 5. Run the test prompt
+
+Now source the local `venv` and run the script using the AWS name of your new endpoint from pulumi state outputs:
```
```suggestion
Now source the local `venv` and run the script using the name of your new endpoint from Pulumi state outputs:
```

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+
+### 6. Cleanup / Teardown
+
+Finally, be sure to clean up un-used resources when you're done with one easy command.
```
```suggestion
Finally, when you're finished with testing your Hugging Face model on SageMaker, you can easily clean up un-used resources with one easy command.
```

scottslowe commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+
+To recap, in a few commands, we created a new Pulumi project from a ready to roll template, deployed an LLM Model endpoint on AWS SageMaker, and tested it with a short python script to generate a response from our model!
+
+AI and ML is becoming more of a necessity every day. It may appear daunting or out of reach at first glance, but with the power of IaC written in Pulumi Python programs, getting started has never been easier. 
```
This is the kind of high-level justification I mentioned having earlier in the post. Maybe move this up there, or copy it up there and recap it here as a reminder?

aaronkao replied:
well this justification is still too nebulous. i would like to see the exact details and customer pain

scottslowe reviewed (changes requested):
A few minor changes, most with suggestions to make it easy to commit. Some general notes:

* We should make sure we always properly capitalize Pulumi and Python (and other languages in general).
* I flagged a couple instances of "AWS SageMaker" but we'll need to make sure all instances are changed to "Amazon SageMaker".

Otherwise, great work! I love to see the results of all the hard work you've been putting in!

aaronkao commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
````diff
@@ -0,0 +1,131 @@
+Finally, be sure to clean up un-used resources when you're done with one easy command.
+
+```bash
+pulumi destroy
````
```suggestion
pulumi destroy -rf --remove
```

i normally do that so they can just nuke the whole thing in one command

usrbinkat replied:
no, this approach is great in a gist or developer docs, but not when we are handholding a new-to-iac audience.

aaronkao commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
````diff
@@ -0,0 +1,131 @@
+
+```bash
+pulumi up
+```
````
can we show a quick image or just text of what the pulumi up output looks like?

elistaykova replied:
+1. That was my first reaction after the reading the post. So what? I think a lot (most) of the code you are showing can be in the readme. Instead, show the interesting infrastructure bits and the resources, the output, what/how was created. Instead of all the details where and how to login

usrbinkat replied:
done

aaronkao commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/meta.png:
Need a new meta image. Remove the hashtag

scottslowe commented:
@usrbinkat I took the liberty of pushing some changes to the branch; apologies for not checking with you first. Please feel free to make sure the added text properly reflects _your_ voice and style. Most of the changes are minor text changes (Amazon SageMaker throughout, fixing capitalization here and there, etc.), but I did take a stab at addressing feedback from Eli and Aaron on providing some context/reason why the article is important/useful/applicable. I would love your touch on this section as well!

usrbinkat commented:
> @usrbinkat I took the liberty of pushing some changes to the branch; apologies for not checking with you first. Please feel free to make sure the added text properly reflects _your_ voice and style. Most of the changes are minor text changes (Amazon SageMaker throughout, fixing capitalization here and there, etc.), but I did take a stab at addressing feedback from Eli and Aaron on providing some context/reason why the article is important/useful/applicable. I would love your touch on this section as well!

I was under the impression we were wanting to s/SageMaker/Sagemaker but double checking AWS branding it's SageMaker, which way are we going with it?
![image](https://github.com/pulumi/pulumi-hugo/assets/54747391/33f90ede-2816-4770-a04e-6fae8099a68d)

scottslowe commented:
> I was under the impression we were wanting to s/SageMaker/Sagemaker but double checking AWS branding it's SageMaker, which way are we going with it?

@usrbinkat We should follow the Amazon branding, for sure.

scottslowe commented:
@usrbinkat Aside from the "Sagemaker versus SageMaker" stuff (we should align with Amazon branding), this is looking really good. I'll defer to Eli and Aaron to review the changes at the top of the article, but otherwise the rest of the article looks good to go.

aaronkao commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,183 @@
+
+In this short tutorial we will deploy a publicly available [Meta AI LlaMa 2] based model from [Hugging Face], on [Amazon SageMaker]. Then we will test it with a [natural language prompt] using a short [Python] script.
+
+It may be tempting to dismiss AI as yet another over hyped and complicated technology, but is it really? What if you could start using this cutting edge technology without the stress? AI/ML models, including [Large Language Models] (LLM) like what we will demonstrate today, offer [tremendous value] to organizations, but only if AI/ML platform resources can be reliably, easily, and cost effectivel ...
```
```suggestion
Running Hugging Face models on Amazon SageMaker is a popular option for deploying AI/ML model backends. However, much of the infrastructure setup, configuration, and deployment is done manually via the SageMaker console. This leads to a number of issues such as manual errors or can your configuration be repeated easily for many deployment. Infrastructure as Code allows you to define all your infrastructure configurations as piece of code that can be version controlled and then the provisioning can be automated. With Pulumi, you can use Python which you are using for everything else to write IaC programs. 
```

aaronkao commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,183 @@
+
+It may be tempting to dismiss AI as yet another over hyped and complicated technology, but is it really? What if you could start using this cutting edge technology without the stress? AI/ML models, including [Large Language Models] (LLM) like what we will demonstrate today, offer [tremendous value] to organizations, but only if AI/ML platform resources can be reliably, easily, and cost effectivel ...
+
+1. Easily create AI/ML infrastructure, together with other supporting infrastructure _in the same Pulumi program_
```
after the above para, doesn't seem this is that necessary

aaronkao commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,183 @@
+
+For this exercise we will deploy a [Pulumi Template] from scratch. Starting from the `pulumi new` command, we will initialize a new [Pulumi project and stack]. This will provide everything required to deploy our chosen Hugging Face model as an [Amazon SageMaker Model Endpoint]
+
+You can use the `sagemaker-aws-python` template as a working Python starting point for your own Amazon SageMaker deployments, and customize the model you use, SageMaker configuration, and CloudWatch integration as well, just to scratch the surface. Your only limits are your imagination.
```
combine this paragraph with the above. 

"For this exercise we will deploy the sagemaker-aws-python template as a working Python starting point.... You will use the pulumi new command to launch this template. This will launch X, Y, Z. Use it as is or as a starting point for bigger infrastructure configurations...

aaronkao commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,183 @@
+
+Now let's look at what the `sagemaker-aws-python` template will deploy in this tutorial:
+
+* [AWS IAM Roles]
```
I would do this maybe in pargraph form so there is more a relationship between everything. 

The template will deploy a Sagemaker instance. IaM roles are attached so it can do <what>. Then huggingface models are downloaded and then mapped to the instance. CW alarms are added to prevent <over cost charges???>. Then an endpoint is exposed and given to you in the outputs. The end point is serving the model

ericrudder commented on themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md:
```diff
@@ -0,0 +1,131 @@
+---
+date: 2023-09-08
+title: "Deploy a Huggingface Model on AWS SageMaker with Pulumi Python IaC"
```
Hugging Face is actually two words (throughout)
